### DISCUSSION TITLE
Windows config paths & themes

### DISCUSSION THREAD

**Discussion by simsrw73** [sharkdp/bat#2428] [Q&A] [answered]
I'm having a lot of trouble on config with Windows (W11 Pro).

One of the first issues I had is that bat doesn't seem to find my config file unless I specifically set "BAT_CONFIG_PATH". I have $HOME and all the $XDG vars set on my Windows system and most x-platform tools honor those variables when set on Windows. It would be a nice-to-have if it honored those vars, but it is resolved by setting "BAT_CONFIG_PATH".

What is unresolved is everything else having to do with config paths. Nothing I do seems to change `bat --config-dir` and using that directory for installing custom themes does not work.

```
❯ bat --config-dir
C:\Users\simsrw73\AppData\Roaming\bat

❯ bat cache --build
No themes were found in 'C:\Users\simsrw73\AppData\Roaming\bat\themes', using the default set
No syntaxes were found in 'C:\Users\simsrw73\AppData\Roaming\bat\syntaxes', using the default set.
Writing theme set to C:\Users\simsrw73\AppData\Local\bat\themes.bin ... okay
Writing syntax set to C:\Users\simsrw73\AppData\Local\bat\syntaxes.bin ... okay
Writing metadata to folder C:\Users\simsrw73\AppData\Local\bat ... okay
```

I've copied the themes into both directories: C:\Users\simsrw73\.config\bat && C:\Users\simsrw73\AppData\Roaming\bat
(Themes are from https://github.com/catppuccin/bat)
```
❯ tree C:\Users\simsrw73\AppData\Roaming\bat

├── bat.conf
└── themes
   ├── Catppuccin-frappe.tmTheme
   ├── Catppuccin-latte.tmTheme
   ├── Catppuccin-macchiato.tmTheme
   └── Catppuccin-mocha.tmTheme
```

I would prefer bat to honor $XDG, so I can keep all my config in my home directory so I can manage it with chezmoi/git. Am I doing something wrong with my config to set locations? Am I doing something wrong to set the theme?

My environment
```
BAT_CONFIG_PATH                 C:\Users\simsrw73\.config\bat\bat.conf
HOME                            C:\Users\simsrw73
HOMEDRIVE                       C:
HOMEPATH                        \Users\simsrw73
PAGER                           moar
USERNAME                        simsrw73
USERPROFILE                     C:\Users\simsrw73
XDG_CACHE_HOME                  C:\Users\simsrw73\.cache
XDG_CONFIG_HOME                 C:\Users\simsrw73\.config
XDG_DATA_HOME                   C:\Users\simsrw73\.local\share
XDG_STATE_HOME                  C:\Users\simsrw73\.local\state
```

❯ bat --diagnostics
#### Software version

bat 0.22.1 (e5d9579)

#### Operating system

Windows 6.2.9200

#### Command-line

```bash
C:\Users\simsrw73\scoop\apps\bat\current\bat.exe --diagnostics
```

#### Environment variables

```bash
SHELL=<not set>
PAGER=moar
LESS=<not set>
LANG=<not set>
LC_ALL=<not set>
BAT_PAGER=<not set>
BAT_CACHE_PATH=<not set>
BAT_CONFIG_PATH=C:\Users\simsrw73\.config\bat\bat.conf
BAT_OPTS=<not set>
BAT_STYLE=<not set>
BAT_TABS=<not set>
BAT_THEME=<not set>
XDG_CONFIG_HOME=C:\Users\simsrw73\.config
XDG_CACHE_HOME=C:\Users\simsrw73\.cache
COLORTERM=<not set>
NO_COLOR=<not set>
MANPAGER=<not set>
```

#### System Config file

Could not read contents of 'C:\ProgramData\bat\config': The system cannot find the path specified. (os error 3).

#### Config file

```
# Set the theme to "TwoDark"
--theme="Coldark-Dark"

# Show line numbers, Git modifications and file header (but no grid)
--style="header,numbers,changes"

# Use italic text on the terminal (not supported on all terminals)
--italic-text=always

# Use C++ syntax for Arduino .ino files
--map-syntax "*.ino:C++"

# Use moar for pager
--pager "moar  -no-linenumbers"
```

#### Custom assets metadata

```
---
bat_version: 0.22.1
creation_time:
  secs_since_epoch: 1672147619
  nanos_since_epoch: 672749100
```

#### Custom assets

- metadata.yaml, 101 bytes
- syntaxes.bin, 898115 bytes
- themes.bin, 40498 bytes

#### Compile time information

- Profile: release
- Target triple: x86_64-pc-windows-msvc
- Family: windows
- OS: windows
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-pc-windows-msvc

#### Less version

```
> C:\Users\simsrw73\scoop\shims\moar.exe --version
v1.11.1
```

**Comment by keith-hall** [comment, marked as answer]
It looks like `bat` is using a crate called [`dirs_next`](https://crates.io/crates/dirs-next) to get config directories:
https://github.com/sharkdp/bat/blob/5519f9c7168d403c2e1c6ad7724854ff5552361b/src/bin/bat/directories.rs#L31-L32

which doesn't seem to respect any of the `XDG`  env vars on Windows. I see that `bat` special cases MacOS, so maybe something similar can be done on Windows too, but the tricky part will presumably be to do so without breaking anyone's current setup.

----

As to why it says
> No themes were found in 'C:\Users\simsrw73\AppData\Roaming\bat\themes', using the default set

I have no idea, if there are `.tmTheme` files there when running `bat cache --build` because it looks like it would only display that if the folder doesn't exist:
https://github.com/sharkdp/bat/blob/a3ea7982466941d746a09bafc8bf386711c48686/src/assets/build_assets.rs#L55-L60

**Reply by simsrw73** [reply]
I was going to try to build this to figure out what is going on, but I'm not really familiar with Rust and was unable to get it to clone (error: invalid path 'Preferences/Symbol List: Blocks.tmPreferences' \nfatal: Unable to checkout '45355633d17ee562481ca1d2ad0e3502c238f58a' in submodule path 'assets/syntaxes/02_Extra/Jinja2') or build (ld: cannot find -lpowrprof) successfully on Windows. Which was disappointing as I wanted to play with Rust a bit. Haven't tried to resolve those as I've got enough demons to chase already.

However, I stumbled on the problem. Which is exactly what it said and you pointed out. The directory didn't exist. If I'd used the command line to copy the files where they needed to go I would not have had any problem. But I've been testing a new File Browser (Files) and... Looking at it now... I still don't understand what it's doing. It shows the directory there when it clearly is not. Sorry for the noise. 

I would still love to have support for storing all the config, data, and cache in $XDG* though if feasible.

![Files](https://user-images.githubusercontent.com/12434447/209837193-16a01733-7a03-48f3-91c8-646ff1e9c805.png)
